<a href="https://colab.research.google.com/github/AlejandroAgduelo22/Business-Analytics/blob/main/Modelo_Arbol_Decision.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**MODELO ÁRBOL DE DECISIÓN RETO 2**

0. Se cargan las librerias

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

#Traemos el modelo
from  sklearn.tree import DecisionTreeClassifier #Esto es el prestamo interbancario


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


1. Cargamos los datos desde el drive

In [ ]:
nxl='/content/drive/MyDrive/Colab Notebooks/2. BD2_In_App_Pur_Int.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XDB=XDB.dropna()  #Evitamos celdas vacias
XDB.head()

#Seleccionamos las variables que nos interesan
XD=XDB[['Age','Income','Years','Hours']]
XD.head()
XD=np.array(XD)                     #Convierto la tabla en números - Variables de entrada
yd=XDB['Buy']; yd=np.array(yd)  #Variable de salida - Convierto la tabla




2. Se procede con la construcció del modelo

In [ ]:
from functools import cmp_to_key
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,yd)
ydp=mar.predict(XD) #Estos valores de pronóstico sirven para evaluar el modelo

#Evaluamos el modelo - Modelos por encima del 75% cumplen
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
VN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
VP=cm[1,1]

Ex=(VN+VP)/(VN+FP+FN+VP)
Se=VP/(VP+FN)
Es=VN/(VN+FP)

print('Exactitud:',Ex)
print('Sensibilidad:',Se)
print('Especificidad:',Es)

Exactitud: 0.8475
Sensibilidad: 0.9478672985781991
Especificidad: 0.7354497354497355


3. Se procede con la visualización del árbol

In [ ]:
from sklearn.tree import export_graphviz       #Estructura del Árbol
from pydotplus import graph_from_dot_data      #Que datos le voy a colocar al árbol

vs=['Age','Income','Years','Hours'] #Los nombres que aparecen en el árbol

dot_data=export_graphviz(mar,feature_names=vs) #Asegurar nombres del gráfico
graph=graph_from_dot_data(dot_data)            #Se hace el gráfico del Árbol
graph.write_png('arbol_Reto2.png')                   #Se crea el árbol

True

4. Le calculamos los valores de Buy a los solicitantes nuevos


In [ ]:
XDB2=pd.read_excel(nxl,sheet_name=1)
XDB2=XDB2[['Age','Income','Years','Hours']]  #Seleccionamos las variables que nos interesan
XDB2=XDB2.dropna()        #Evita las celdad vacias

XD2=np.array(XDB2)        #Convierto la tabla en números - Variables de Entrada
ydp2=mar.predict(XD2)     #Estos valores de pronóstico sirven para evaluar el modelo
XDB2['Buy']=ydp2       #Le pego la columna a la base de datos
XDB2.head()
XDB2.to_excel('NuevosCompradoresR2.xlsx')

np.sum(ydp2)
print(ydp2)

[1 0]


En la base de datos se puede observar un total de 400 posibles compradores en el juego en 189 posibles no compradores (47%) y un total de 211 que si comprarían (53%). De acuerdo con lo anterior, el modelo de árbol arrojo un indice de exactitud cercano al 85%, así como indices de Sensibilidad cercanos al 95% y una especificidad del 74% que muestran el comportamiento del modelo frente a la clasificación de posibles compradores. Este mismo porcentaje en promedio se pudo observar frente al indice de especifidad el cual mide el comportamiento del modelo en la clasificación de posibles compradores. Es importante mencionar que un modelo que exhiba valores de exactitud (Accuarcy) por encima del 75%, es considerado un modelo aceptable para el pronóstico de los datos. Para la base de datos de posibles compradres nuevos (2 compradores), se puede observar que el modelo clasifico un total de 1 en no comprador (50%) en la categoría de compradores y un total de 0 en compradores (50%) en la categoría de compradores. Lo anterior muestra la predominancia del modelo por la clasificación de la categoría de no compradores.